exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/v3/schema/ScanSchemaTracker.java

```diff
@@ -0,0 +1,466 @@
+   * The result set loader provides a dummy reader for unprojected columns.
+   * (A reader can, via the result set loader, find if a column is projected if
+   * doing so helps reader efficiency.)
+   * <p>
+   * The projection filter is the first line of defense for schema conflicts.
+   * The {code ResultSetLoader} will query the filter with a full column
+   * schema. If that schema conflicts with the scan schema for that column,
+   * this method will raise a {@code UserException}, which typically indicates
+   * a programming error, or a very odd data source in which a column changes
+   * types between batches.
+   *
+   * @param errorContext the reader-specific error context to use if
+   * errors are found
+   */
+  ProjectionFilter projectionFilter(CustomErrorContext errorContext);
+
+  /**
+   * Once a reader has read a batch, the reader will have provided a type
+   * for each projected column which the reader knows about. For a wildcard
+   * projection, the reader will have added all the columns that it found.
+   * This call takes the reader output schema and merges it with the current
+   * scan schema to resolve dynamic types to concrete types and to add
+   * newly discovered columns.
+   * <p>
+   * The process can raise an exception if the reader projects a column that
+   * it shouldn't (which is not actually possible because of the way the
+   * {@code ResultSetLoader} works.) An error can also occur if the reader
+   * provides a type different than that already defined in the scan schema
+   * by a defined schema, a provided schema, or a previous reader in the same
+   * scan. In such cases, the reader is expected to have converted its input
+   * type to the specified type, which was presumably selected because the
+   * reader is capable of the required conversion.
+   *
+   * @param readerOutputSchema the actual schema produced by a reader when
+   * reading a record batch
+   * @param errorContext the reader-specific error context to use if
+   * errors are found
+   */
+  void applyReaderSchema(TupleMetadata readerOutputSchema,
+      CustomErrorContext errorContext);
+
+  /**
+   * The missing column handler obtains the list of missing columns from
+   * {@link #missingColumns()}. Depending on the scan lifecycle, some of the
+   * columns may have a type, others may be dynamic. The missing column handler
+   * chooses a type for any dynamic columns, then calls this method to tell
+   * the scan schema tracker the now-resolved column type.
+   * <p>
+   * Note: a goal of the provided/defined schema system is to avoid the need
+   * to guess types for missing columns since doing so quite often leads
+   * to problems further downstream in the query. Ideally, the type of missing
+   * columns will be known (via the provided or defined schema) to avoid
+   * such conflicts.
+   */
+  void resolveMissingCols(TupleMetadata missingCols);
+
+  CustomErrorContext errorContext();
+  MutableTupleMetadata schema();
```

Comment:
Maybe call this mutableSchema? So the caller knows exactly what type of schema is given.